Migrate pkg to slog

benoittgt opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:

tl;dr : migrate packages and cmd from `log` to `slog` for logs

While trying to implement https://github.com/helm/helm/pull/30689 but for V4. It was [mentionned](https://github.com/helm/helm/pull/30685#discussion_r2006477698) that I should use `slog`. So I tried to make a first try in 
https://github.com/helm/helm/pull/30685, but it requires a change accross packages and cmd.

This PR migrates packages and cmd to `slog`.

- Migrate all log to `slog` key-value, with everything in Debug, except logs that start with `warning:`
- Start all logs phrase with lowercase letter [like recommend in Go](https://github.com/helm/helm/pull/30603#discussion_r1980279824)
- Try to make logs more easily grepable by adding more often the name and the namespace of the resource
- Follow same construction in struct as before, so SDK's users can pass custom slogger with a custom handler to process logs

This code still need to be fully manually tested but first I would like to get some feedbacks.

cc @robertsirc @TerryHowe 

Close: https://github.com/helm/helm/pull/30698

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

benoittgt committed:
Migrate kube package to slog

As for helm v4. We want to migrate logs to slog.

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Remove unreachable error

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Do no mask warning alerts

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Try to make log more common and more easily grepable

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Provide an adapter to easily pass a slog.Default()

```
helmClient.Log = NewSlogAdapter(slog.Default())
```

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Extra comment + Default logger fallback

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Some interesting rephrasing by Terry Howe

See: https://github.com/helm/helm/pull/30698#discussion_r2012394228
Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Properly use DefaultLogger

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Deal with linting errors

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Fix call to kube log

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt commented on pkg/kube/logger.go:
```diff
@@ -0,0 +1,66 @@
+package kube
+
+import "log/slog"
+
```
I am wondering where this should be, because I don't think it should be limited to `kube` package. Internal package?

robertsirc replied:
not sure what you mean?

benoittgt replied:
The issue I see is because interface for all helm packages use this slog adapter from `kube`, it will create unwanted links to kube package. I made a commit to show the issue (for me): https://github.com/helm/helm/commit/a1ca89b72a3ca45c81a6362dfd4bb6f122ade1ff#diff-a15c00615f24e3d585c1e7dee030d66e0e1d1666a34ce1ff038a84dbe2f5dba5L99

When I see [this](https://github.com/helm/helm/commit/a1ca89b72a3ca45c81a6362dfd4bb6f122ade1ff#diff-326ded87050c2e687ffc462ccc58f298102a4c3234645f356d1bd7350bda82d8R32), I don't think this is something we want.

I am wondering if
1. We want a slog adapter to still be able to pass a custom logger in packages ? 🙇🏻‍♂️
2. We want to have the slog adapter as an internal package like [`urlutil`](https://github.com/helm/helm/tree/main/internal/urlutil) ?

robertsirc replied:
Sorry for the late delay let me take a deeper look at it and we can bring this up at the next Helm developer meeting.

benoittgt committed:
Demonstrate the impact of having Logger defined in kube package

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Migrate to a dedicated internal package for slog adapter + migrate more

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Migrate more code to log adapter

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Fix linting issue and temporary removing logging in test acion

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Fix missing logger to SQL in test

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Explain why we ignore the timestamp

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Migrate to pure slog without a custom wrapper

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Properly discard by default logs

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Enforce error style with others

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Do not remove the functionality to print log in test

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt commented on pkg/cmd/root.go:
```diff
@@ -95,16 +95,7 @@ By default, the default directories depend on the Operating System. The defaults
-func Warning(format string, v ...interface{}) {
-	fmt.Fprintf(os.Stderr, "WARNING: "+format+"\n", v...)
-}
+var Logger = cli.NewLogger(settings.Debug)
```
Not sure if this is what we want with how messages are rendered with slog [TextHandler](https://pkg.go.dev/golang.org/x/exp/slog#TextHandler). We have very few `key=value` with `Logger.Debug` or `Logger.Warn`, so I think it's ok.

benoittgt replied:
As with the last commits. https://github.com/helm/helm/pull/30708/commits/e7eedae97cb8ea9050ec220a7cb33445c0b791f9

```
benoit.tigeot » bin/helm upgrade --install --debug --wait frontend \         
--namespace test \
--set replicaCount=2 \
--set backend=http://backend-podinfo:9898/echo \
podinfo/podinfo
level=DEBUG msg="getting history for release" release=frontend
level=DEBUG msg="getting release history" name=frontend
level=DEBUG msg="preparing upgrade" name=frontend
```

benoittgt committed:
Properly reproduce the nopLogger as before

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Trick slog to return the full error

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Go the slog way

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Linting

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Prevent redefining verbose flags

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Leverage slog.Any for errors

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

robertsirc commented on internal/monocular/client.go:
```diff
@@ -44,12 +45,10 @@ func New(u string) (*Client, error) {
 	return &Client{
 		BaseURL: u,
-		Log:     nopLogger,
+		Log:     slog.New(slog.NewTextHandler(io.Discard, nil)),
```
I think we want to remove this logger that is passed around and address logging directly from slog.

robertsirc reviewed (changes requested):
First let me thank you for all your hard work on this. Where we are calling logging looks good but how we implement the logging i think needs to change. Rather than passing around a logger `Log` as to just call `slog` directly. Would love to setup some time to disucss if you are up to it.

benoittgt committed:
debug log level is dynamic and set after Logger creation

So we should use dynamic handler to set the log level after. With this
patch we can clearly see the output. Before we were always stuck in log
level "info" and not seeing debug log level

```
bin/helm upgrade --install --debug --wait frontend \
--namespace test \
--set replicaCount=2 \
--set backend=http://backend-podinfo:9898/echo \
podinfo/podinfo
level=DEBUG msg="getting history for release" release=frontend
level=DEBUG msg="preparing upgrade" name=frontend
level=DEBUG msg="performing update" name=frontend
level=DEBUG msg="creating upgraded release" name=frontend
level=DEBUG msg="checking resources for changes" resources=2
level=DEBUG msg="no changes detected" kind=Service name=frontend-podinfo
level=DEBUG msg="patching resource" kind=Deployment name=frontend-podinfo namespace=test
level=DEBUG msg="waiting for resources" count=2 timeout=5m0s
level=DEBUG msg="waiting for resource" name=frontend-podinfo kind=Deployment expectedStatus=Current actualStatus=Unknown
level=DEBUG msg="updating status for upgraded release" name=frontend
Release "frontend" has been upgraded. Happy Helming!
NAME: frontend
LAST DEPLOYED: Thu Apr 10 09:56:25 2025
NAMESPACE: test
STATUS: deployed
REVISION: 6
DESCRIPTION: Upgrade complete
```

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt commented on pkg/cli/logger.go:
```diff
@@ -0,0 +1,43 @@
+	}
+
+	// Create a handler that removes timestamps
+	handler := slog.NewTextHandler(os.Stdout, &slog.HandlerOptions{
```
I thought we wanted to chose extactly the output, like do not have the timestamp. Also to handle debug loglevel like it was done before it seems we need to have a custom slog `Handler` that handle the debug flag just before log is processed.

benoittgt commented:
> Would love to setup some time to disucss if you are up to it. 
I will ping you on Slack. ;)

benoittgt committed:
Call slog directly instead of using a wrapper

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Use the logger with proper handling of dynamic debug on 2 locations

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
No longer log call location using flag, it will need to be done in slog

Some ideas here: https://www.reddit.com/r/golang/comments/15nwnkl/achieve_lshortfile_with_slog/

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt commented:
After a quick call with @robertsirc I removed all mention of `Log` and instead leveral `slog` directly + `slog.SetDefault`.

1. SDK developer will still be able to set a custom default logger with a dedicated handler thank to `SetDefault`
2. I choose to make [a `Handler` for slog](https://github.com/helm/helm/pull/30708/files#diff-d7242d95f7b1b021fe7ed2365df3ea1b95065e4da55dccf9ec5098b6304ea4f8) to do not loose the `--debug` behavior.
3. We loose the call location with the logging with slog 7f02e89a7a3c88120c424f46471dd11ac9650db5
, but I think log message are pretty uniq and it's easy to find the location of the call with basic text search on Github.
4. We may want to improve errors. 
    ```
    benoit.tigeot » bin/helm uninstall --namespace test podinfo 

    Error: uninstall: Release not loaded: podinfo: release: not found

    benoit.tigeot » bin/helm uninstall --namespace test podinfo --debug

    level=DEBUG msg="getting release history" name=podinfo
    Error: uninstall: Release not loaded: podinfo: release: not found
    level=DEBUG msg=error error="release: not found\nhelm.sh/helm/v4/pkg/storage/driver.init\n\t<autogenerated>:1\nruntime.doInit1\n\truntime/proc.go:7291\nruntime.doInit\n\truntime/proc.go:7258\nruntime.main\n\truntime/proc.go:254\nruntime.goexit\n\truntime/asm_arm64.s:1223\nuninstall: Release not loaded: podinfo\nhelm.sh/helm/v4/pkg/action.(*Uninstall).Run\n\thelm.sh/helm/v4/pkg/action/uninstall.go:87\nhelm.sh/helm/v4/pkg/cmd.newUninstallCmd.func2\n\thelm.sh/helm/v4/pkg/cmd/uninstall.go:60\ngithub.com/spf13/cobra.(*Command).execute\n\tgithub.com/spf13/cobra@v1.9.1/command.go:1015\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\tgithub.com/spf13/cobra@v1.9.1/command.go:1148\ngithub.com/spf13/cobra.(*Command).Execute\n\tgithub.com/spf13/cobra@v1.9.1/command.go:1071\nmain.main\n\thelm.sh/helm/v4/cmd/helm/helm.go:48\nruntime.main\n\truntime/proc.go:272\nruntime.goexit\n\truntime/asm_arm64.s:1223"
    ```

Sorry again for the size of the PR.

benoittgt committed:
Fix nil pointer dereference in ready test

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Prefer using slog.Any when displaying errors

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

robertsirc reviewed (approved):
LGTM, there are some issues that we are going to want to follow up with another PR to fix but since this is part of Helm4, we have some liberties.

mattfarina commented on pkg/kube/client.go:
```diff
@@ -719,9 +715,6 @@ func copyRequestStreamToWriter(request *rest.Request, podName, containerName str
 	if err != nil {
 		return errors.Errorf("Failed to copy IO from logs for pod: %s, container: %s", podName, containerName)
 	}
-	if err != nil {
-		return errors.Errorf("Failed to close reader for pod: %s, container: %s", podName, containerName)
-	}
```
Note, this was unreachable code so it's safe to remove.
